Update provider so the `useI18n()` hook can use the global translation functions from `@wordpress/i18n`

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

`@automattic/react-i18n` works by providing locale data somewhere near the root of your component tree (using `<I18nProvider>`), and using the `useI18n()` hook in your components to get access to translation functions that use the provided locale data.

However we don't always have access to the locale data. In the ETK, core initialises `@wordpress/i18n` with the locale data and we don't get access to it. That means `__` from `@wordpress/i18n` works, but `__` from `useI18n()` does not.

This PR ~~adds a `<GlobalI18nProvider>` provider that is useful~~ changes the `<I18nProvider>` provider so it is useful in situations where we don't have access to the raw locale data.  When this provider is used, `useI18n()` will provide access to the global translation functions from `@wordpress/i18n`.

~~You still need to provide the current locale to `<GlobalI18nProvider>` somehow. You may also need to provide the current locale to `<I18nProvider>` somehow. That's because `@wordpress/i18n` doesn't give access to what the current locale is for some reason and `react-i18n` does.~~ (doing #47446 instead)

In this PR I've tested it out by translating a few strings in the launch flow ~~and I've gotten the current locale by having the server render it to a global. I previously thought I'd be able to grab the locale slug from `document.documentElement.lang` but unfortunately it doesn't always match the locale slug (e.g. sometimes the locale is `fr` and the `<html lang="...">` attribute is `fr_CA`)~~

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply D52644-code to sandbox
* Sandbox an unlaunched gutenboarding site
* Ensure language is something other than `en`
* Open the block editor for your test site and click the launch button in the top right
* The 4 steps in the sidebar should still be translated
  ![Screen Shot on 2020-11-12 at 18:24:39](https://user-images.githubusercontent.com/1500769/98899387-5a549b80-2514-11eb-802e-80defaa0ea9c.png)

p-jackson commented:
@sirreal I think you were involved at the start of `@automattic/react-i18n`, could you take a look to see whether this seems like an ok approach? Also do you happen to have any insights into why `@wordpress/i18n` doesn't expose the current locale?

jsnajdr commented:
This looks like an OK solution that solves a product issue. But I wish we could refactor it to something that doesn't need to introduce a new provider. The existing one should be able to work even with the default instance of `I18n`:
```js
import { defaultI18n } from '@wordpress/data';

<I18nProvider localeData={ defaultI18n.getLocaleData() }>
```

To make that happen, the `@wordpress/data` library should export the `defaultI18n` object, not only a subset of its bound methods like `__`. And it should get a new `getLocaleData()` getter method.

> do you happen to have any insights into why `@wordpress/i18n` doesn't expose the current locale?

I'm not sure it's guaranteed that the library always _knows_ the locale 🙂 All it needs to work is a `localeData` object that maps English strings to other strings. The information about what the "other" language actually is is optional. The `localeData` objects have a special `''` key with some metadata about the language:
```js
{
  "": {
    "plural_forms": "nplurals=3; plural=(n==1) ? 0 : (n>=2 && n<=4) ? 1 : 2;",
    "language": "cs_CZ",
    "localeSlug":"cs"
  }
}
```
But I don't know if it's 100% guaranteed in practice that this special key will always be there. Without it, the library will incorrectly format plurals, because it doesn't have information about the target language's grammar and will use English rules as default. But other than that, it will work.

Once the `getLocaleData()` getter is exposed, the `localeSlug` can be accessed as
```js
i18n.getLocaleData()[ '' ].localeSlug
```
or, preferrably, there can be convenience getters for the `language` and `localeSlug` values.

All these would be a valuable enhancements to `@wordpress/data` that in the end remove the need for an extra `<GlobalI18nProvider>`.

`i18n-calypso` already has these methods and adding them to `@wordpress/data` would lead to further convergence between both libraries.

yuliyan commented:
> Once the `getLocaleData()` getter is exposed, the `localeSlug` can be accessed as
> 
> ```js
> i18n.getLocaleData()[ '' ].localeSlug
> ```
> 
> or, preferrably, there can be convenience getters for the `language` and `localeSlug` values.

I like the suggested approach, but want to add that `language`, `localeSlug` and `localeVariant` should not be considered as standardized properties. We have them being added on WP.com with the internal i18n deploy process, but it's not guaranteed that these will exist on Jed json files generated from elsewhere. For example, GlotPress' Jed files include only `lang` property.

p-jackson commented:
> The existing one should be able to work even with the default instance of `I18n`

This would be great! My original idea was something like `<I18nProvider i18n={ defaultI18n }>` until I realised that GB wasn't exposing the default `I18n` instance.

> To make that happen, the `@wordpress/data` library should export the `defaultI18n` object, not only a subset of its bound methods like `__`.

Just to clarify @jsnajdr, did you mean the `@wordpress/i18n` package? Or do you think the locale data needs to be exposed via a Redux store? Personally I think just having more access to `@wordpress/i18n` is enough and we don't need to move things to `@wordpress/data`.

> But I don't know if it's 100% guaranteed in practice that this special key will always be there

Thanks, good to know. We're probably relying on the locale slug too much. We use it to add `?locale=es` query params to some API calls, and while that's needed in some places, for some of the APIs we're calling we should probably just add support for `?locale=user`. `localizeUrl()` does assume that the locale will be known on the client.

> `language`, `localeSlug` and `localeVariant` should not be considered as standardized properties. We have them being added on WP.com with the internal i18n deploy process ...

So it sounds like it's ok for our a8c specific functions (like `localizeUrl()`) to use these properties, but anything that we expect to work in core shouldn't refer to these properties at all.

jsnajdr commented:
> Just to clarify @jsnajdr, did you mean the @wordpress/i18n package?

Yes, I always meant `@wordpress/i18n` and the `data` package has absolutely nothing to do with this 🙂 I fixed my original comment now.

lsl commented on packages/react-i18n/src/index.tsx:
```diff
@@ -27,6 +27,28 @@ export const I18nProvider: React.FunctionComponent< Props > = ( { children, loca
+			_nx,
+			_x,
+			isRTL,
+			i18nLocale: localeSlug,
```
Can you add `sprintf` here too? - it doesn't make sense to leave it out

p-jackson replied:
The `I18n` interface doesn’t include `sprintf` because it’s a dumb formatting utility, it doesn’t need access to any locale data. I suppose it might be convenient, but it doesn’t feel right to me to provide something with a hook that can be gotten from an `import`.

lsl replied:
The `@wordpress/i18n` api does include sprintf though, and thats what we're trying to mirror / expose / augment here as a hook, that's why see it as worth including, adding it all into the hook makes usage more consistent and easier to explain / demonstrate.

> it doesn’t feel right to me to provide something with a hook that can be gotten from an import.

Agree in general, this situation for me is more about providing a consistent experience when using l10n functions. I would even argue it isn't worth exposing by wordpress/i18n at all - its just a simple wrapper around another library - but I suspect it was included for this consistency reason to ensure everyone had all the tools in a one stop shop sort of way.

lsl replied:
For the record, don't mind too much either way, go with your gut

```
import { useI18n, sprintf } from '@automattic/react-i18n'
const { __, _x } = useI18n()
```

vs

```
import { useI18n } from '@automattic/react-i18n'
const { __, _x, sprintf } = useI18n()
```

p-jackson commented:
The `I18n` interface doesn’t include `sprintf` because it’s a dumb formatting utility, it doesn’t need access to any locale data. I suppose it might be convenient, but it doesn’t feel right to me to provide something with a hook that can be gotten from an `import`.

> On 16/11/2020, at 1:17 PM, Louis Laugesen <notifications@github.com> wrote:
> 
> 
> @lsl commented on this pull request.
> 
> In packages/react-i18n/src/index.tsx <https://github.com/Automattic/wp-calypso/pull/47352#discussion_r523836775>:
> 
> > +interface GlobalI18nProviderProps {
> +	localeSlug: string;
> +}
> +
> +export const GlobalI18nProvider: React.FunctionComponent< GlobalI18nProviderProps > = ( {
> +	children,
> +	localeSlug,
> +} ) => {
> +	const contextValue = React.useMemo< I18nReact >(
> +		() => ( {
> +			__,
> +			_n,
> +			_nx,
> +			_x,
> +			isRTL,
> +			i18nLocale: localeSlug,
> Can you add sprintf here too? - it doesn't make sense to leave it out
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub <https://github.com/Automattic/wp-calypso/pull/47352#pullrequestreview-530853156>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AALOMYM5ZRT2N4RSRCPRE7LSQBVRVANCNFSM4TSZZWTQ>.
>

p-jackson committed:
I18nProvider uses global __ function if localeData is left out

p-jackson committed:
Launch sidebar uses react-i18n for some of it's translations

p-jackson commented:
@lsl What do you think about these changes? I've removed `<GlobalI18nProvider>` and now if the `<I18nProvider>` component is used without a `localeData` prop it will default to using the global translation functions.

lsl commented on packages/react-i18n/src/index.tsx:
```diff
@@ -65,12 +71,19 @@ export const withI18n = createHigherOrderComponent< I18nReact >( ( InnerComponen
  */
-function makeContextValue( localeData?: LocaleData ): I18nReact {
+function makeContextValue( localeData?: LocaleData, localeSlug?: string ): I18nReact {
+	const i18nLocale = localeSlug ?? localeData?.[ '' ]?.localeSlug ?? 'en';
```
Neither accessing localeSlug or the `en` fallback work here, localeSlug is non standard and en fallback can't be assumed otherwise.

lsl replied:
(the en fallback above is fine, I just don't think it makes sense in this package)

p-jackson replied:
Indeed. The `useI18n()` hook guarantees that it'll always return a value for `i18nLocale`. However maybe that's not realistic. Perhaps we should change the return type of `useI18n()` to this:

```typescript
export interface I18nReact {
	__: I18n[ '__' ];
	_n: I18n[ '_n' ];
	_nx: I18n[ '_nx' ];
	_x: I18n[ '_x' ];
	isRTL: I18n[ 'isRTL' ];
	i18nLocale?: string;
}
```

Note the **?** on the `i18nLocale` prop. That way we'd fall back to `undefined` instead of `'en'`.

**Another** option I've been thinking about is completely removing the `i18nLocale` prop from `useI18n()`. `react-i18n` is supposed to be something we could push upstream, and since having the locale slug available in the locale data is non-standard, it probably won't survive the review process.

We've been explicitly referring to the locale slug a lot in gutenboarding, but that's a special case because the user is logged out, and they can provide the locale in the url. That's not usually how things work. Perhaps the locale slug can be provided to a *new* thing, like a `<LocaleProvider>`, that's part of an a8c-specific package. `i18n-utils` I guess. That way we can depend on it for our a8c helper functions like `localizeUrl()` and we know we can provide it at app boot time because of the idiosyncrasies of our WordPress environment.

lsl replied:
I don't think an optional locale slug value works - users of the prop will have to deal with that case everywhere its used, the default value needs to be decided by the consumer of react-i18n. In a lot of cases that will be english but its an assumption we can't force on everyone.

This another other option sounds really good - feels much cleaner and I like stepping away from making decisions about what a locale is.

lsl replied:
#47446 - I like this solution to break apart the locale slug from the locale data, but we need some changes to break away from `calypso-i18n` https://github.com/Automattic/wp-calypso/pull/47446#pullrequestreview-530931752

cc @Automattic/i18n might want to be aware of this

p-jackson replied:
I think we should go with #47446. It probably needs to merge first so this one can be updated without breaking the build.

p-jackson commented:
As of #47446 the `<I18nProvider>` should no longer takes a `localeSlug` prop. We'll see which PR lands first!

lsl reviewed (dismissed):
Per https://github.com/Automattic/wp-calypso/pull/47352#discussion_r525900533

Requesting changes to block this PR until https://github.com/Automattic/wp-calypso/pull/47446 is merged.

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -69,8 +69,14 @@ export const withI18n = createHigherOrderComponent< I18nReact >( ( InnerComponen
 	const i18nLocale = localeData?.[ '' ]?.localeSlug ?? 'en';
+
+	if ( ! localeData ) {
+		return { __, _n, _nx, _x, isRTL, i18nLocale };
```
As discussed on a P2, the `i18nLocale`'s default value `en` will be wrong if the upstream global `i18n` instance is actually localized.

A good alternative would be to not pass the `i18nLocale` prop in this case at all, or completely remove it from the context value.

p-jackson replied:
I'm going to remove `i18nLocale` since currently there's currently nothing depending on it. Not ruling out that it won't eventually make it's way back in some form though.

jsnajdr commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/attach-launch-sidebar.tsx:
```diff
@@ -41,9 +42,11 @@ registerPlugin( 'a8c-editor-site-launch', {
-			<LaunchContext.Provider value={ { siteId: window._currentSiteId } }>
-				<LaunchModal onClose={ closeSidebar } />
-			</LaunchContext.Provider>
+			<I18nProvider>
```
You don't even need to mount the provider here: the default context value, derived from `undefined` `localeData`, will be used in that case.

p-jackson replied:
Good thinking, I'd somehow missed this in the midst of all these changes. I actually like this behaviour 😎

jsnajdr reviewed (approved):
Looks good :+1: Doing something about the `i18nLocale` being always `en` would be desirable, but it's up to you to decide what to do and when. Thanks for improving the `react-i18n` library!

p-jackson committed:
Remove i18nLocale from react-i18n

There's no way to know the correct locale slug when `localeData` is left
undefined, so just removing it since there's currently no uses of it.

The current locale slug might come back to react-i18n in the future once
there's a standard way to get access to the current locale slug.
